[REF] delivery_carrier_label_postlogistics becomes delivery_postlogistics in 13.0

phuctranfxvn opened the pull request:
Description:
- New module `delivery_postlogistics` as a refactoring of module `delivery_carrier_label_postlogistics` to get it closer to the native `delivery` module

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,231 @@
+    _inherit = "delivery.carrier"
+
+    # TODO: recheck if we need this code
+    code = fields.Char()
```
To remove

phuctranfxvn replied:
done

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,231 @@
+    postlogistics_endpoint_url = fields.Char(
+        string="Endpoint URL", default="https://wedecint.post.ch/", required=True,
+    )
+    postlogistics_client_id = fields.Char(string="Client ID")
```
Add groups="base.group_system"

phuctranfxvn replied:
fixed

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,231 @@
+        string="Endpoint URL", default="https://wedecint.post.ch/", required=True,
+    )
+    postlogistics_client_id = fields.Char(string="Client ID")
+    postlogistics_client_secret = fields.Char(string="Client Secret")
```
Add groups="base.group_system"

phuctranfxvn replied:
fixed

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,231 @@
+            if carrier.prod_environment:
+                carrier.postlogistics_endpoint_url = "https://wedec.post.ch/"
+            else:
+                carrier.postlogistics_endpoint_url = "https://wedec.post.ch/"
```
Wrong URL according to above description

phuctranfxvn replied:
fixed

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,231 @@
+                "warning_message": False,
+            }
+        else:
+            return {
```
Delivery can be free of charge to the end customer. This should not return an error.
Postlogistics has no api for delivery price rating?
In this case, it should always return success

phuctranfxvn replied:
this function is from the label module, not currently used. Removed

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,231 @@
+
+    def postlogistics_get_tracking_link(self, picking):
+        return (
+            "https://service.post.ch/EasyTrack/"
```
No different endpoint for test?

simahawk replied:
@phuctranerp what about the question by @jbaoudoux ?

jbaudoux commented on delivery_postlogistics/models/postlogistics_carrier_template_options.py:
```diff
@@ -0,0 +1,57 @@
+    name = fields.Char(translate=True)
+    code = fields.Char()
+    description = fields.Char(
+        help="Allow to define a more complete description " "than in the name field.",
```
```suggestion
        help="Allow to define a more complete description than in the name field.",
```

phuctranfxvn replied:
fixed

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,287 @@
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    delivery_type = fields.Selection(related="carrier_id.delivery_type")
```
Already defined in module delivery, do not redefine

jbaudoux replied:
this probably relates to https://github.com/OCA/delivery-carrier/pull/323/files#diff-7d5856a39a59766b6cd8d580e267bd8b19f48ff5ceccb17d7243e90632feef21R38 but is this issue still valid for v13 ?

phuctranfxvn replied:
removed in the latest update

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,287 @@
+    _inherit = "stock.picking"
+
+    delivery_type = fields.Selection(related="carrier_id.delivery_type")
+    carrier_code = fields.Char(related="carrier_id.code")
```
Seems not used, so to remove

phuctranfxvn replied:
removed

jbaudoux commented on delivery_postlogistics/views/stock.xml:
```diff
@@ -0,0 +1,41 @@
+                    type="object"
+                />
+            </field>
+            <!-- <xpath expr="//page//field[@name='carrier_id']" position="after">
```
useless

phuctranfxvn replied:
removed

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,287 @@
+        comodel_name="postlogistics.delivery.carrier.option", string="Options"
+    )
+
+    delivery_fixed_date = fields.Date(
```
This should somehow be linked with the sales order commitment_date when set

phuctranfxvn replied:
yep, it should be, added

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,287 @@
+    delivery_type = fields.Selection(related="carrier_id.delivery_type")
+    carrier_code = fields.Char(related="carrier_id.code")
+    # TODO: Change the name of this field
+    option_ids = fields.Many2many(
```
prefix with postlogistics

phuctranfxvn replied:
done

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,477 @@
+        if partner.street2:
+            recipient["addressSuffix"] = partner.street2
+
+        if partner.parent_id and partner.parent_id.name != partner_name:
```
This should go up the tree structure until the parent that is flagged as is_company and not the first parent

phuctranfxvn replied:
fixed

jbaudoux reviewed (commented):
First code review, not tested yet

jbaudoux commented on delivery_postlogistics/views/stock.xml:
```diff
@@ -0,0 +1,41 @@
+        <field name="inherit_id" ref="delivery.view_picking_withcarrier_out_form" />
+        <field name="arch" type="xml">
+            <field name="state" position="before">
+                <button
```
This action already exists in standard and is called "send to shipper". This calls "postlogistics_send_shipping"

phuctranfxvn replied:
fixed, removed this button and the action will be called in `def send_to_shipper()`

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,286 @@
+                package = self.env["stock.quant.package"].create({})
+                move_lines.write({"result_package_id": package.id})
+
+    def action_generate_carrier_label(self):
```
This should be the method "postlogistics_send_shipping"

phuctranfxvn replied:
refactored

jbaudoux reviewed (changes requested):
Method postlogistics_send_shipping must be properly implemented. Please move content of action_generate_carrier_label into that method

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,286 @@
+                # triggered the onchange:
+                # https://github.com/odoo/odoo/issues/2693#issuecomment-56825399
+                # Ideally we should add the missing option
+                raise exceptions.UserError(
```
This prevents to change the carrier as this trigger this onchange. If postlogictics has a mandatory option, then this error will raise.

jbaudoux replied:
Can you dig this issue. Configure postlogistics with a mandatory option. Take a picking without carrier and set the postlogistic carrier. This triggers the onchange on carrier that will fill the mandatory option as defined here https://github.com/OCA/delivery-carrier/pull/323/files#diff-7d5856a39a59766b6cd8d580e267bd8b19f48ff5ceccb17d7243e90632feef21R46
then this will trigger this onchange but the error should not be raised as the option as already been added.

Highcooley replied:
The error will also raise, when a new SO is created with this carrier or if the missing mandatory option is tried to be add to an existing SO with this carrier.

phuctranfxvn replied:
the field `postlogistics_option_ids` on picking will be removed soon, so we will remove this method also

phuctranfxvn replied:
fields postlogistics_option_ids and the function related to this field removed in the latest update

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,490 @@
+        if "ZAW3219" in option_codes and picking.delivery_place:
+            attributes["deliveryPlace"] = picking.delivery_place
+        if picking.carrier_id.postlogistics_proclima_logo:
+            attributes["proClima"] = True
```
Seems to be true by default (meaning that if you don't provide a value, the logo is printed). So I would always send this value

phuctranfxvn replied:
I have fixed this, adding the case for False value

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,286 @@
+            ):
+                # XXX the client does not allow to modify the field that
+                # triggered the onchange:
+                # https://github.com/odoo/odoo/issues/2693#issuecomment-56825399
```
Not sure this is still valid and nor useful to keep

phuctranfxvn replied:
this method will be removed when we remove the field on picking which will be replaced by the product packages assigned to the picking

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+        """
+        partner = picking.partner_id
+
+        partner_name = partner.name or partner.parent_id.name
```
Must be improved to include company in name2.
Note to myself: provide the right specs

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+        if partner.street2:
+            recipient["addressSuffix"] = partner.street2
+
+        company_partner = partner
```
company_partner = partner.commercial_company_name

No need to recompute it

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+        if "ZAW3218" in option_codes and pack_num:
+            attributes.update(
+                {
+                    "parcelTotal": pack_total or picking.number_of_packages,
```
I cannot find where is defined picking.number_of_packages. I do not find it in any module.

phuctranfxvn replied:
this field is available in Odoo 11, removed from odoo 12, do you think that we can use the length of the field `package_ids` ?

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+            }
+            if carrier.postlogistics_tracking_format == "picking_num":
+                if not package:
+                    # start with 9 to garentee uniqueness and use 7 digits
```
```suggestion
                    # start with 9 to ensure uniqueness and use 7 digits
```

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+        def add_item(package=None):
+            assert picking or package
+            itemid = self._get_itemid(
+                picking, package.name if package else picking.name
```
_get_itemid concatenate the picking.name and second argument. So, if there is no package, it will display twice the picking.name. I would change with "else None" + see the remark above in _get_itemid

phuctranfxvn replied:
fixed

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+
+        """
+        name = _compile_itemid.sub("", picking.name)
+        if pack_no:
```
if no pack_no, return the name

phuctranfxvn replied:
fixed

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,287 @@
+                "name": tracking_number + "." + label["file_type"],
+            }
+
+        labels = []
```
Can you extract the writing of the tracking numbers in another method?

phuctranfxvn replied:
done

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,492 @@
+        #     franking_license = company_licenses[0]
+        return franking_license.number
+
+    def _prepare_attributes(self, picking, pack_num=None, pack_total=None):
```
Add pack in the attributes
self, picking, pack=None, pack_num=0, pack_total=0

Regarding PRZL (service), we need to be able to choose it when creating the package.
For this, there is the concept of product.packaging that is defined in the delivery module. When a package is created a product.packaging is assigned to it. So, we need to load a list of product.packaging for postlogistics with all the possible service values. Then we need to be able to define a default postlogistics product packaging (i.e. default service) on the carrier.
If there is no package, the value PRZL is the carrier default service. Else for each package, is it is the service defined on the assigned product packaging.
This also means that the service will not be defined by an option code anymore.

phuctranfxvn replied:
@jbaudoux I dont really understand this. Do you mean that we can use the `product.packaging` generated when creating a package for defining the services przl ?
What is the default carrier service ? For the assigning product.packaging, which values should be used ? `Package Code` I think

jbaudoux replied:
All the product.packaging possibilities must be pre-loaded with a data file and not created during processing. This defines the service that is used.
At SO, the post carrier is chosen and on that carrier, for instance, it defaults to PRI service (default product.packaging on the carrier). Then at delivery when you create the package, you can still select a product.packaging (for instance SP service due to over sized package). So you assign that product.packaging to the package (if you do not assign a specific one, it defaults to the default value of the carrier). Based on the product.packaging, you know what is the service to declare.
DHL enterprise module is working similarly.

phuctranfxvn replied:
Understand, will update the source code for this soon. Thanks

phuctranfxvn replied:
@jbaudoux as far as I check, each move of the picking will be assigned to a package (stock.quant.package) and a product packaging type (product.packaging). Therefore if we want to define multiple services for each picking, we may need to define multiple services in product packaging level. Correct me if I am wrong.
Btw, when I do the test with PRI (only one service: PRI), I got this error message.
```
Swisspost API returns errors:
Error code: E2030
Error message: Basic service PostPac Priority does not belong to an international service group.
```
Do you have any idea?

jbaudoux replied:
@phuctranerp Indeed, each package (stock.quant.package) has a product packaging that has a defined service. So each package has indirectly a service associated.

Regarding the error, maybe in your test you didn't used a from and to address in Switzerland?

phuctranfxvn replied:
I have adjusted the address to have the package delivered internally in Switzerland. However, then I get the error 500 - Internal Server Error. Not sure why. Here is the json data I posted
URL: https://wedecint.post.ch/api/barcode/v1/generateAddressLabel
Json Data
```
{
    "language": "DE",
    "frankingLicense": "60013217",
    "customer": {
        "name1": "A Swiss Company",
        "street": "Rue de Lausanne 1",
        "zip": "1030",
        "city": "Bussigny",
        "country": "CH"
    },
    "labelDefinition": {
        "labelLayout": "A7",
        "printAddresses": "RECIPIENT_AND_CUSTOMER",
        "imageFileType": "PNG",
        "imageResolution": "200"
    },
    "item": {
        "itemID": "WHOUT00005+PACK0000005",
        "recipient": {
            "name1": "Wood Corner",
            "street": "EPFL Innovation Park, B\u00e2t A",
            "zip": "1015",
            "city": "Lausanne",
            "email": "wood.corner26@example.com",
            "country": "CH"
        },
        "attributes": {
            "przl": [
                "PRI"
            ],
            "weight": 0.16
        }
    }
}
```

I will recheck it again, it could be an issue of testing server, or do you have any idea? @jbaudoux ?
previously, I tested with this instance and it worked well

phuctranfxvn replied:
works well now by changing the end point url to https://wedec.post.ch/
done for delivery service on product.packaging

jbaudoux replied:
This is still not good when you have multiple packs. From what I understand and as I wrote initially, you need to know for which pack you are generating the attributes: self, picking, **pack=None**, pack_num=0, pack_total=0

In case of multiple packs, picking._get_picking_postlogistic_packaging() is not returning the product packaging of the processed pack.


Then regarding combined DLC like for instance "SP, PRI" for "Bulky goods Priority", as there is only one shipper_package_code on the product packaging, we could define it as "SP, PRI" but then you need here to split it to declare the service as ["SP","PRI"]

phuctranfxvn replied:
Hi, sorry I missed this comment. Thanks for your recommendation, will be update soon.

phuctranfxvn replied:
fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555

jbaudoux reviewed (changes requested):
Some additional remarks

TDu commented:
I only see one commit, but there seems to be an history from the list of authors.
Have we lost the commit history for this ?

TDu commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,294 @@
+        if self.delivery_type == "postlogistics":
+            self.postlogistics_send_shipping()
+
+    def postlogistics_send_shipping(self):
```
Shouldn't this implementation be in `delivery.carrier` model so it works with 
https://github.com/odoo/odoo/blob/6b363d202d9c35e8249f7bb2b0730b995fdf4a7d/addons/delivery/models/delivery_carrier.py#L173
?

TDu replied:
Oh, actually this function is missing :
https://github.com/OCA/delivery-carrier/blob/e518473604ec38eb416420fe0fb38227960b0f5d/delivery_carrier_label_postlogistics/models/delivery.py#L224

phuctranfxvn replied:
Hi @TDu , you are right, actually the method implemented in picking object should be moved to delivery.carrier so that the the method will be executed when the method send_to_shipper is executed.
This will be fixed, I am discussing with jbaudoux to make some changes services & options to be implemented, and then this will be included in next update.

phuctranfxvn replied:
fixed with the latest commit updates

yvaucher commented:
@phuctranerp Can you describe in the description of this Pull request what is the intent of it?

Is it a a new module or a migration of `delivery_carrier_label_postlogistics`.
I see that most code comes from the latest.

As it looks like to be a migration you will want to keep the git history.

nilshamerlinck commented:
Hi @TDu and @yvaucher 

`delivery_postlogistics` is a refactoring of `delivery_carrier_label_postlogistics` with the intent of bringing it closer to the native `delivery` framework

As such, and as discussed with @jbaudoux, the plan for next steps is to:
- close https://github.com/OCA/delivery-carrier/pull/318 that is rendered useless
- explicitly include the commits history from `delivery_carrier_label_postlogistics` in this PR indeed
- fix pending issues being discussed

TDu commented:
Are we not missing this specific method rate_shipment method which was implemented in 12.0 
https://github.com/OCA/delivery-carrier/blob/e518473604ec38eb416420fe0fb38227960b0f5d/delivery_carrier_label_postlogistics/models/delivery.py#L203

phuctranfxvn commented:
> Are we not missing this specific method rate_shipment method which was implemented in 12.0
> https://github.com/OCA/delivery-carrier/blob/e518473604ec38eb416420fe0fb38227960b0f5d/delivery_carrier_label_postlogistics/models/delivery.py#L203

Hi @TDu, thanks for reviewing, you are right, added back

phuctranfxvn commented:
> @phuctranerp Can you describe in the description of this Pull request what is the intent of it?
> 
> Is it a a new module or a migration of `delivery_carrier_label_postlogistics`.
> I see that most code comes from the latest.
> 
> As it looks like to be a migration you will want to keep the git history.

Hi @yvaucher
Thanks for your feedback, this has been taken into account

jbaudoux commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,145 @@
+    def postlogistics_rate_shipment(self, order):
+        self.ensure_one()
+        delivery_product_price = self.product_id and self.product_id.lst_price
+        if delivery_product_price:
```
As commented previously here https://github.com/OCA/delivery-carrier/pull/323#discussion_r540949991
Please always return success as free of charge should not be handled as an error (but do not remove the method)

phuctranfxvn replied:
Fixed

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,499 @@
+
+        attributes = {
+            "przl": services,
+            "weight": picking.shipping_weight,
```
1/ Should be the weight of the pack.
2/ It is in grams in the API while in kg in Odoo. So you have to multiply by 1000

phuctranfxvn replied:
Fixed.
I introduced another commit that supports the multiple packages in the same picking.
I just reviewed the source code and I see there is a bug in the original postlogistics module that no matter how many package in the picking, only one package will be pushed. Please help me review and let me know

jbaudoux reviewed (changes requested):
Still some issues regarding the attributes

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,516 @@
+        services = []
+        if package:
+            postlogistics_packagings = package.mapped("packaging_id")
+            services = [
```
packaging = package.packaging_id or picking.carrier_id.postlogistics_default_packaging_id
services = packaging.shipper_package_code and package.packaging_id.shipper_package_code.split(",") or []

phuctranfxvn replied:
updated, thanks for your feedback.

jbaudoux commented on delivery_postlogistics/views/product_packaging_view.xml:
```diff
@@ -0,0 +1,19 @@
+        <field name="arch" type="xml">
+            <field name="shipper_package_code" position="after">
+                <field
+                    name="postlogistics_option_code"
```
Should not be required

phuctranfxvn replied:
updated, thanks for your feedback.

jbaudoux reviewed (changes requested):
Can you please not "force-push" otherwise it's very complicate to track your changes. Thanks

phuctranfxvn commented:
> Can you please not "force-push" otherwise it's very complicate to track your changes. Thanks

well noted, thanks

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -202,21 +195,18 @@ def _prepare_attributes(
-            packaging.postlogistics_option_code
-            for packaging in postlogistics_packagings
-        ]
+        option_code = packaging.postlogistics_option_code
```
With your previous change, service codes (DLC) are now recorded in the field shipper_package_code, so we don't need anymore the field postlogistics_option_code. Many services (=options) can be activated.

phuctranfxvn replied:
fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -202,21 +195,18 @@ def _prepare_attributes(
+        option_code = packaging.postlogistics_option_code
 
-        if "ZAW3217" in option_codes and picking.delivery_fixed_date:
+        if option_code == "ZAW3217" and picking.delivery_fixed_date:
```
If this option is activate, we can only enable it if there is a fixed date.
```
if "ZAW3217" in services:
  if picking.delivery_fixed_date:
    attributes["deliveryDate"] = picking.delivery_fixed_date
  else:
    services.remove("ZAW3217")
```

phuctranfxvn replied:
fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -202,21 +195,18 @@ def _prepare_attributes(
+        if option_code == "ZAW3217" and picking.delivery_fixed_date:
             attributes["deliveryDate"] = picking.delivery_fixed_date
-        if "ZAW3218" in option_codes and pack_num:
+        if option_code == "ZAW3218" and pack_num:
```
ZAW3218 is not related to parcel number.

Replace with:
`if pack_num:`

phuctranfxvn replied:
fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -202,21 +195,18 @@ def _prepare_attributes(
             "przl": services,
```
To move at the end of the method as some activated services should be removed (see below)

phuctranfxvn replied:
fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555

jbaudoux reviewed (changes requested):
Some clarifications regarding the services and options after re-reading the API doc

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,506 @@
+        is_phone_required = [
+            packaging
+            for packaging in picking_packagings
+            if packaging.postlogistics_option_code == "ZAW3213"
```
This option should be defined on the partner, not on the packaging.

phuctranfxvn replied:
well noted, to be updated

phuctranfxvn replied:
Fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/f04b053b0aeca3c51756f5ad9129349ce26f4555

jbaudoux reviewed (commented):
Regarding the services (options), some are packaging related (priority, fragile, ...), some are partner related (notify by phone), some are SO data related (commitment date), some are product related (hazarouds goods)

It would make sense to be able to activate options also on a partner and when building the PRZL attribute, combine them all.

jbaudoux commented:
@phuctranerp Could you merge this PR https://github.com/trobz/delivery-carrier/pull/1 ? Thanks

phuctranfxvn commented:
> @phuctranerp Could you merge this PR [trobz#1](https://github.com/trobz/delivery-carrier/pull/1) ? Thanks

Merged, thanks for your PR

yvaucher reviewed (changes requested):
@phuctranerp please follow OCA directives to port a module. You need to keep the history of the module even if it has been renamed.

https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#howto

```
$ git clone https://github.com/OCA/$REPO -b 13.0
$ git checkout -b 13.0-mig-$MODULE origin/13.0
$ git format-patch --keep-subject --stdout origin/13.0..origin/12.0 -- $MODULE | git am -3 --keep
$ pre-commit run -a  # to run black, isort and prettier (ignore pylint errors at this stage)
$ git add -A
$ git commit -m "[IMP] $MODULE: black, isort, prettier"  --no-verify  # it is important to do all formatting in one commit the first time
```

where in you case the module name is delivery_carrier_label_postlogistics

phuctranfxvn commented:
> $ git checkout -b 13.0-mig-$MODULE origin/13.0



> @phuctranerp please follow OCA directives to port a module. You need to keep the history of the module even if it has been renamed.
> 
> https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-13.0#howto
> 
> ```
> $ git clone https://github.com/OCA/$REPO -b 13.0
> $ git checkout -b 13.0-mig-$MODULE origin/13.0
> $ git format-patch --keep-subject --stdout origin/13.0..origin/12.0 -- $MODULE | git am -3 --keep
> $ pre-commit run -a  # to run black, isort and prettier (ignore pylint errors at this stage)
> $ git add -A
> $ git commit -m "[IMP] $MODULE: black, isort, prettier"  --no-verify  # it is important to do all formatting in one commit the first time
> ```
> 
> where in you case the module name is delivery_carrier_label_postlogistics

Hi @yvaucher

Let me know if I missed something in my PR. Actually I included the commits of the old module already.

yvaucher commented:
@phuctranerp weird, those are indeed in the commit tab but not anymore on the conversation tab. I guess the Github layout has changed recently.

How did you proceed to get the commits? it seems the date of the commits were reset.


Otherwise you can squash commits for translations.
like
https://github.com/OCA/delivery-carrier/pull/323/commits/a93c499736b066e5f71cc3e755a026560100f22e
https://github.com/OCA/delivery-carrier/pull/323/commits/5bddc4255d94a8ff4f7b4755abb72887fe4d5d5b
https://github.com/OCA/delivery-carrier/pull/323/commits/07a1754d4f9ce6eef73ddd41e0c26e85c24836ed

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,497 @@
+            )
+
+        attributes = {
+            "weight": total_weight,
```
Weight must be an integer and not a float
Weight cannot exceed 99999 otherwise an exception must be raised
This should also be fixed in the tests

phuctranfxvn replied:
@jbaudoux, about the max weight, last time I check, the error will be raised when the weight exceed 30kg (which is 30,000g). Btw, I dont think we need to put a validation here, because when we push the package, and if there is an error raised from the server, our odoo will raise an exception for that (weight exceeding should be validated in server side, in case the max value is changed by postlogistics).
For the type of the weight, I will force-format it in integer type, updated in https://github.com/OCA/delivery-carrier/pull/323/commits/3b74e94c38f331ff5ff54cb2e1c0525991379f44

jbaudoux replied:
@phuctranerp ok. Thanks

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,497 @@
+        # get options
+        if not user_lang:
+            user_lang = "en_US"
+        lang = self._get_language(user_lang)
```
The language must be the language of the partner not of the user. Seems reasonable to remove from the `generate_label` method signature and pass here as argument `picking.partner_id.lang`

phuctranfxvn replied:
Updated in https://github.com/OCA/delivery-carrier/pull/323/commits/3b74e94c38f331ff5ff54cb2e1c0525991379f44

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,497 @@
+            "zip": partner.zip,
+            "city": partner.city,
+            "country": partner.country_id.code,
+            "domicilePostOffice": picking.carrier_id.postlogistics_office,
```
When not used, must be `None` instead of `False`. This seems the rule for the strings other "False" is printed on the label
```suggestion
            "domicilePostOffice": picking.carrier_id.postlogistics_office or None,
```

phuctranfxvn replied:
updated in https://github.com/OCA/delivery-carrier/pull/323/commits/3b74e94c38f331ff5ff54cb2e1c0525991379f44

jbaudoux reviewed (changes requested):
Some remarks received from the swiss post.

phuctranfxvn commented:
> @phuctranerp weird, those are indeed in the commit tab but not anymore on the conversation tab. I guess the Github layout has changed recently.
> 
> How did you proceed to get the commits? it seems the date of the commits were reset.
> 
> Otherwise you can squash commits for translations.
> like
> [a93c499](https://github.com/OCA/delivery-carrier/commit/a93c499736b066e5f71cc3e755a026560100f22e)
> [5bddc42](https://github.com/OCA/delivery-carrier/commit/5bddc4255d94a8ff4f7b4755abb72887fe4d5d5b)
> [07a1754](https://github.com/OCA/delivery-carrier/commit/07a1754d4f9ce6eef73ddd41e0c26e85c24836ed)

As normal process, I checkout a new branch from 13.0, using git format-patch for pushing the modules, then replace the module with the new one. I dont know why the date of the commits were reset also.

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,495 @@
+        if pack_num:
+            attributes.update(
+                {
+                    "parcelTotal": pack_total or len(picking.package_ids),
```
Parcel counter starts at 0
```suggestion
                    "parcelTotal": (pack_total or len(picking.package_ids)) - 1,
```

phuctranfxvn replied:
Thanks for suggestion, fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/3cf38c4e77e36768bbdff33cdb8ff33d7b56093c

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,495 @@
+            attributes.update(
+                {
+                    "parcelTotal": pack_total or len(picking.package_ids),
+                    "parcelNo": pack_num,
```
```suggestion
                    "parcelNo": pack_num - 1,
```

phuctranfxvn replied:
Thanks for suggestion, fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/3cf38c4e77e36768bbdff33cdb8ff33d7b56093c

jbaudoux reviewed (changes requested):
Parcel counter starts at 0

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,495 @@
+            "street": partner.street,
+            "zip": partner.zip,
+            "city": partner.city,
+            "email": partner.email or None,
```
Email notification is not free of charge. Can you add a boolean on the partner to enable/disable notification. By default, notification must be disabled

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,495 @@
+            "street": partner.street,
+            "zip": partner.zip,
+            "city": partner.city,
+            "email": partner.email or None,
```
Email notification is not free of charge.
Can you add a "postlogistics notification" selection on the partner with the options:
* disabled
* email
* sms (mobile)
* phone call

By default, notification must be disabled

Only provide email if notification on partner is "email"

phuctranfxvn replied:
Thanks for suggestion, fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/3cf38c4e77e36768bbdff33cdb8ff33d7b56093c

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,495 @@
+
+        # Phone and / or mobile should only be displayed if instruction to
+        # Notify delivery by telephone is set
+        postlogistics_partner_codes = partner.postlogistics_option_ids.mapped("code")
```
Replace this by "postlogistics notification" selection.
If notification by "sms", then send mobile number.
If notification by "phone call", then enable "ZAW3213" and set is_phone_required

phuctranfxvn replied:
Thanks for suggestion, fixed in https://github.com/OCA/delivery-carrier/pull/323/commits/3cf38c4e77e36768bbdff33cdb8ff33d7b56093c

jbaudoux reviewed (changes requested):
We also need some improvements regarding the notifications. There are still ongoing clarification with the Swiss Post, but with those changes we should have enough to test

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+            else:
+                services.remove("ZAW3217")
+
+        if pack_num:
```
After clarification, parcelNo/Total cannot be used if service is not activated
```
if "ZAW3218" in services:
  if pack_total > 1:
    ...
  else:
     services.remove("ZAW3218")
```

phuctranfxvn replied:
Added in this commit https://github.com/OCA/delivery-carrier/pull/323/commits/1cc0ca72a0f173d2d7c3f63dc43ca868ea3e48f5

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+            "street": partner.street,
+            "zip": partner.zip,
+            "city": partner.city,
+            "email": partner.postlogistics_notification == "email"
```
Just for clarity, can you move it down like you did for the other?
```
if partner.postlogistics_notification == "email":
  recipient["email"] = partner.email
elif partner.postlogistics_notification == "sms":
  recipient["mobile"] = partner_mobile
elif partner.postlogistics_notification == "phone":
  ...      
```

phuctranfxvn replied:
Added in this commit https://github.com/OCA/delivery-carrier/pull/323/commits/1cc0ca72a0f173d2d7c3f63dc43ca868ea3e48f5

jbaudoux commented on delivery_postlogistics/models/res_partner.py:
```diff
@@ -0,0 +1,26 @@
+
+    postlogistics_notification = fields.Selection(
+        [
+            ("disabled", "Disable"),
```
```suggestion
            ("disabled", "Disabled"),
```

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+
+        partner_name = partner.name or partner.parent_id.name
+        recipient = {
+            "name1": partner_name,
```
```suggestion
            "name1": partner_name[:35],
```

phuctranfxvn replied:
Added in https://github.com/OCA/delivery-carrier/pull/323/commits/29e751d5d5865b752a91ef4195bc06ae167d944f, thanks

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+        partner_name = partner.name or partner.parent_id.name
+        recipient = {
+            "name1": partner_name,
+            "street": partner.street,
```
```suggestion
            "street": partner.street[35],
```

phuctranfxvn replied:
Added in https://github.com/OCA/delivery-carrier/pull/323/commits/29e751d5d5865b752a91ef4195bc06ae167d944f, thanks

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+            "name1": partner_name,
+            "street": partner.street,
+            "zip": partner.zip,
+            "city": partner.city,
```
```suggestion
            "city": partner.city[:35],
```

phuctranfxvn replied:
Added in https://github.com/OCA/delivery-carrier/pull/323/commits/29e751d5d5865b752a91ef4195bc06ae167d944f, thanks

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+        recipient = {
+            "name1": partner_name,
+            "street": partner.street,
+            "zip": partner.zip,
```
```suggestion
            "zip": partner.zip[:10],
```

phuctranfxvn replied:
Added in https://github.com/OCA/delivery-carrier/pull/323/commits/29e751d5d5865b752a91ef4195bc06ae167d944f, thanks

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+            recipient["country"] = partner.country_id.code.upper()
+
+        if partner.street2:
+            recipient["addressSuffix"] = partner.street2
```
```suggestion
            recipient["addressSuffix"] = partner.street2[:35]
```

phuctranfxvn replied:
Added in https://github.com/OCA/delivery-carrier/pull/323/commits/29e751d5d5865b752a91ef4195bc06ae167d944f, thanks

jbaudoux reviewed (changes requested):
Recipient data cannot exceed a defined length

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+        partner_name = partner.name or partner.parent_id.name
+        recipient = {
+            "name1": partner_name[:35],
+            "street": partner.street[35],
```
Small typo
```suggestion
            "street": partner.street[:35],
```

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,230 @@
+        def info_from_label(label):
+            tracking_number = label["tracking_number"]
+            return {
+                "file": base64.b64decode(label["binary"]),
```
After base64 decoding, for **ZPL labels only** (not for PDF), you also have to call ".decode('cp437')"
Then the label is missing the enabling of UTF-8 on the printer. So you have to append after "^XA" the code "^CI28"

jbaudoux replied:
For the context, when the city is "Zürich", the received data is encoded in a way that it cannot be printed

jgrandguillaume replied:
I receive a file, if you requested a ZPL -> Call ".decode('cp437')". Then
"^XA" -> "^XA^CI28"

jbaudoux replied:
Pay attention the content of base64.b64decode(label["binary"]) is still in base64. So it requires a second decoding before decoding with the right cp

phuctranfxvn replied:
@jbaudoux, yes, I see it. do you think it was done on purpose ? I actually need to decode twice to get the content

jbaudoux replied:
So it should be sth like:
if label["file_type"] == "zpl2"
  data = base64.b64encode(base64.b64decode(base64.b64decode(label["binary"])).decode('cp437').encode('utf-8').replace('^XA","^XA^CI28"))

santostelmo replied:
https://github.com/trobz/delivery-carrier/pull/2

phuctranfxvn replied:
@santostelmo thanks for your PR, merged

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+
+        company_partner_name = partner.commercial_company_name
+        if company_partner_name and company_partner_name != partner_name:
+            recipient["name2"] = partner.parent_id.name
```
```suggestion
            recipient["name2"] = partner.parent_id.name[:35]
```

jbaudoux commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,510 @@
+
+    def _get_item_additional_data(self, picking, package=None):
+        result = []
+        packaging_codes = package and package.packaging_id._get_packaging_codes() or []
```
When the package has no packaging_id, we should raise an error to the user: "The package %s must have a package type"

**Traceback**
{{ packaging_codes = package and package.packaging_id._get_packaging_codes() or []}}
{{ File "/odoo/external-src/delivery-carrier/delivery_postlogistics/models/product_packaging.py", line 18, in _get_packaging_codes}}
{{ self.ensure_one()}}
{{ File "/odoo/src/odoo/models.py", line 5104, in ensure_one}}
{{ raise ValueError("Expected singleton: %s" % self)}}
ValueError: Expected singleton: product.packaging()

phuctranfxvn replied:
Add in https://github.com/OCA/delivery-carrier/pull/323/commits/430e52d51d740eb55e962336dd11056827aba033

jbaudoux commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,238 @@
+                data = base64.b64encode(
+                    base64.b64decode(data)
+                    .decode("cp437")
+                    .replace("^XA", "^XA^CI28")
```
We apparently also need to enable the right font on the printer.
Could you make this entire replace configurable on the carrier.
If boolean "Do not patch ZPL to enable UTF-8" is False (default), then apply the replace
If True, then do not replace
```suggestion
                    .replace("^XA", "^XA^CW0,E:TT0003M_.TTF^XZ^XA^CI28")
```

jbaudoux replied:
Maybe instead of a boolean on the carrier, we could put the full replacement string. Then it's not hardcoded anymore

phuctranfxvn replied:
Add in https://github.com/OCA/delivery-carrier/pull/323/commits/430e52d51d740eb55e962336dd11056827aba033

jbaudoux replied:
Thanks

simahawk commented:
May I suggest to add the old module to 13.0 as (not installable and disabled in pre-commit) and keep only relevant commits here? It would ease reviews...

jbaudoux commented:
> May I suggest to add the old module to 13.0 as (not installable and disabled in pre-commit) and keep only relevant commits here? It would ease reviews...

@simahawk Why? There is no reason for keeping the old module as it is replaced by this one.

jbaudoux reviewed (approved):
Thanks for all the fixes. According to me, this can now be merged.

simahawk commented:
> > May I suggest to add the old module to 13.0 as (not installable and disabled in pre-commit) and keep only relevant commits here? It would ease reviews...
> 
> @simahawk Why? There is no reason for keeping the old module as it is replaced by this one.

you don't keep it. You do what you have done here to change it. Doesn't change anything in the result. The only thing that changes is that you don't get 108 commits to review, new files instead of moved files, etc.

simahawk commented:
One more thing that is missing in this PR is squashing translation commits as far as possible.
I prepared a base here https://github.com/OCA/delivery-carrier/pull/343
Once merged will leave this PR w/ only a bunch of commits (less than 20 and many squash-able).

simahawk commented:
@phuctranerp can you rebase on 13.0 and drop all commits but the ones for the refactoring? While doing so, please drop all merge commits and squash what you can. Thanks!

simahawk commented:
it seems all translations are gone. Is this wanted?

phuctranfxvn commented:
> @phuctranerp can you rebase on 13.0 and drop all commits but the ones for the refactoring? While doing so, please drop all merge commits and squash what you can. Thanks!

yes sure, done.

simahawk commented:
> > @phuctranerp can you rebase on 13.0 and drop all commits but the ones for the refactoring? While doing so, please drop all merge commits and squash what you can. Thanks!
> 
> yes sure, done.

Thanks! Some remarks:

1. the module is not renamed so only a few files have been detected and deleted or moved. Likely the old module is still there.
2. translations are missing
3. no migration from old version?

phuctranfxvn commented:
> > > @phuctranerp can you rebase on 13.0 and drop all commits but the ones for the refactoring? While doing so, please drop all merge commits and squash what you can. Thanks!
> > 
> > 
> > yes sure, done.
> 
> Thanks! Some remarks:
> 
> 1. the module is not renamed so only a few files have been detected and deleted or moved. Likely the old module is still there.
> 2. translations are missing
> 3. no migration from old version?

Hi @simahawk

Thanks for your feedbacks

1. I have removed the old module in the latest update already
2. As discussed, the translation will be handled by MQT
3. To be updated soon.

UPDATED: I have added the migration script, cc @simahawk

simahawk commented on delivery_postlogistics/migrations/13.0.1.0.0/pre-migration.py:
```diff
@@ -0,0 +1,36 @@
+    # delivery.carrier.template.otion => postlogistics.delivery.carrier.template.option
+    rename_model_list = [
+        (
+            "delivery.carrier.template.option",
```
I'm not sure this is correct. This model was defined `base_delivery_carrier_label` and you could have many *_carrier_label modules installed.
Probably we should simply drop the model and rely on the fact that all other carrier modules will register their own records.
@yvaucher any advice on this?

simahawk commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,242 @@
+    delivery_fixed_date = fields.Date(
+        "Fixed delivery date", help="Specific delivery date (ZAW3217)"
+    )
+    delivery_place = fields.Char(
```
wouldn't be better to have new partner relation here?
But I see you preserve an original behavior. Not mandatory to change.

phuctranfxvn replied:
> @phuctranerp what about the question by @jbaoudoux ?

@simahawk I dont know the correct endpoint for testing instance for this, and I cannot find it in postlogistics' document. However, afaik, environment now depends on the account type, as last time we use the production endpoint for testing. Any idea ?

simahawk replied:
just leave it as this but add `TODO: consider refactoring these fields using a partner relation instead`

phuctranfxvn replied:
updated

simahawk commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,242 @@
+        """ Inherit this method in your module """
+        if not self.carrier_id:
+            return
+        # This can look useless as the field carrier_code and
```
indeed, it looks useless :)
The related field is readonly... I don't understand the purpose of this.

phuctranfxvn replied:
@simahawk value of the related field will not be triggered to update unless the original field is saved to the db. So I think this is a trick to trigger the change of the related field on changing the data of the original field.

The related field decides how the view looks like, so we need it to be updated.
What do you think?

simahawk replied:
If you tell me it works... fine, I learn something new :wink: 
Judging only from code, since `delivery_type` is already related to carrier.delivery_type and is readonly IDT the form is going to trigger an update. But I might be wrong... :man_shrugging:

phuctranfxvn replied:
ah I checked again, I am wrong :D. Yes agree that we don't need this function actually.
Removed

simahawk commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,242 @@
+        for operation in operations:
+            # Take the destination package. If empty, the package is
+            # moved so take the source one.
+            packages |= operation.result_package_id or operation.package_id
```
these can be very slow w/ lots of records. Would you mind adding a `TODO: refactor using set of ids instead of browse records and browse at the end`?

phuctranfxvn replied:
updated

simahawk commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,242 @@
+        Note we can receive multiple labels for a same package
+        """
+
+        def info_from_label(label, zpl_patch_string=False):
```
shouldn't this func stay out of `write_tracking_number_label`?

phuctranfxvn replied:
@simahawk `write_tracking_number_label` is now the only function that calls this method, but yes, agree with you that we should move this function out. To be updated soon

phuctranfxvn replied:
updated

simahawk commented on delivery_postlogistics/models/stock_quant_package.py:
```diff
@@ -0,0 +1,74 @@
+    @api.returns("stock.picking")
+    def _get_origin_pickings(self):
+        self.ensure_one()
+        operation_model = self.env["stock.pack.operation"]
```
:fire: hmm am I wrong or this model does not exists in v13? Shouldn't you look for s.move.line? :thinking: 
If this is true, it means this piece of code is not called by any test. If so, please fix test cov too. :pray:

phuctranfxvn replied:
updated

simahawk commented on delivery_postlogistics/tests/test_postlogistics.py:
```diff
@@ -0,0 +1,126 @@
+class TestPostlogistics(common.SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
```suggestion
        super().setUpClass()
        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```

phuctranfxvn replied:
updated

simahawk commented on delivery_postlogistics/__manifest__.py:
```diff
@@ -0,0 +1,28 @@
+    ],
+    "installable": True,
+    "auto_install": False,
+    "application": True,
```
application?

phuctranfxvn replied:
@simahawk I am not really sure if we should put "application" here, but I found that some `delivery_` modules in enterprise (delivery_ups, delivery_fedex, ...) are defined as applications

simahawk reviewed (approved):
code review only, LG.

@jbaudoux @yvaucher good for you?

simahawk commented:
@phuctranerp thanks for your work and kind collaboration!

/ocabot merge nobump

simahawk commented:
@phuctranerp build is failing on merge branch. Can you check?
It seems VCR is missing which seems weird since tests are green here... Can you check if the are running properly?

phuctranfxvn committed:
[REF] delivery_carrier_label_postlogistics becomes delivery_postlogistics in 13.0

phuctranfxvn commented:
> @phuctranerp build is failing on merge branch. Can you check?
> It seems VCR is missing which seems weird since tests are green here... Can you check if the are running properly?

Hello @simahawk, I have updated the vcr lib. It should be ok now. Thanks for your report

yvaucher commented on delivery_postlogistics/views/delivery.xml:
```diff
@@ -0,0 +1,95 @@
+        parent="delivery.menu_delivery"
+        sequence="20"
+        action="action_postlogistics_delivery_carrier_options"
+    /> -->
```
Options menu to re-activate ?

yvaucher commented on delivery_postlogistics/readme/CONFIGURE.rst:
```diff
@@ -6,8 +6,8 @@
-* Set your login informations
+* Go to `Inventory -> Configuration -> Delivery -> Shipping Methods`
+* Create new shipping methods for postlogistics and set your login informations
 * launch the Update PostLogistics Services
```
It seems the `Update PostLogistics Services` which was there to download the list of option is not there anymore. (might have been removed in previous changes already)

yvaucher commented on delivery_postlogistics/readme/CONFIGURE.rst:
```diff
@@ -6,8 +6,8 @@
-* Go to `Configurations -> Settings -> Postlogistics`
-* Set your login informations
+* Go to `Inventory -> Configuration -> Delivery -> Shipping Methods`
+* Create new shipping methods for postlogistics and set your login informations
```
```suggestion
* Create new shipping methods for PostLogistics and set your login information
```

yvaucher commented on delivery_postlogistics/__manifest__.py:
```diff
@@ -0,0 +1,28 @@
+# © 2013-2016 Yannick Vaucher (Camptocamp SA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Postlogistics Shipping",
```
```suggestion
PostLogistics Shipping - “Barcode” web service
```

https://geschaeftsbericht.post.ch/17/ar/en/category/post_logistics_en/

PostLogistics is written with a capital `L`

And better to name it after the official name so users can find it quickly.
https://www.post.ch/en/post-barcode-handbuch.pdf
https://www.post.ch/en/sending-letters/franking-mail/franking-solutions-for-letters/support-for-consignments-with-barcode/sending-consignments-web-service-barcode

yvaucher commented on delivery_postlogistics/__manifest__.py:
```diff
@@ -0,0 +1,28 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "Postlogistics Shipping",
+    "summary": "Print postlogistics shipping labels",
```
```suggestion
    "summary": "Print PostLogistics shipping labels using the Barcode web service",
```

yvaucher commented on delivery_postlogistics/postlogistics/web_service.py:
```diff
@@ -0,0 +1,520 @@
+
+        if not services:
+            raise exceptions.UserError(
+                _("No Postlogistics packaging services found in this picking.")
```
```suggestion
                _("No PostLogistics packaging services found in packaging {} of picking {}.").format(...)
```

Instead of picking please mention the picking and the packaging, plus put the names in there.
It will greatly help with module that will create labels in batch to be able to identify which packaging has an issue.

We should not assume the label is always generated from the UI of the picking.

We can probably also mention the `Basic services` as it requires at least a Basic service (like ECO, PRI....) and then can have additional services.

yvaucher commented on delivery_postlogistics/models/product_packaging.py:
```diff
@@ -0,0 +1,19 @@
+from odoo import fields, models
+
+
+class ProductPackaging(models.Model):
```
Can we make `shipper_package_code` required when `postlogistics` is selected?

yvaucher commented on delivery_postlogistics/models/product_packaging.py:
```diff
@@ -0,0 +1,19 @@
+    _inherit = "product.packaging"
+
+    package_carrier_type = fields.Selection(
+        selection_add=[("postlogistics", "Postlogistics")]
```
```suggestion
        selection_add=[("postlogistics", "PostLogistics")]
```

yvaucher commented on delivery_postlogistics/models/delivery_carrier.py:
```diff
@@ -0,0 +1,136 @@
+    _inherit = "delivery.carrier"
+
+    delivery_type = fields.Selection(
+        selection_add=[("postlogistics", "Post logistics")]
```
```suggestion
        selection_add=[("postlogistics", "PostLogistics")]
```

sebalix commented:
@phuctranerp are you OK to merge this in your PR? https://github.com/trobz/delivery-carrier/pull/4

yvaucher reviewed (changes requested):
Thanks for the changes, finally had time to dive into this.

I see some great improvements by being closer to the Odoo standard.

I'm concerned though about the removal of the options feature. Which will let the user clueless on what to do to configure it.


With the removal of the dependency on `base_delivery_carrier_label` and the option management I understand the following changes.

1. Options are not pulled anymore from the webservice
2. Options are new to be set on `product.packaging` inside `shipper_package_code` but there is no guidelines anymore whereas we had the list of option available before.
3. As there are additional options, it's not possible to add additional options anymore, we have to select a different `product.packaging`. Which means we have to make almost all possible configuration of packaging? There is no per picking or per package tweaking anymore.


Some of those changes could be problematic as we use this for client that massively use this module that generate labels in batches.



BTW, instructions in `README.rst` still mention

```
* launch the Update PostLogistics Services

This will load available services and generate carrier options.
Now you can create a carrier method for PostLogistics WebService:
* First choose a Service group and save
* Add a Mandatory Carrier option using a Basic Service
* Save Carrier Method (this will update filters to show you only
  compatible services)
* Then add other `Optional as default` and `Optional` carrier option
  from listed
* Add additional Service and Delivery instructions
```

Which is not valid anymore. It should mention the configuration of `product.packaging`
 (to change in USAGE.rst or INSTALL.rst)


I make the following suggestions:

1. Fixup the documentation for the setup to reflect the current state of the module
2. Add a UI text that lists the available options in the UI when `postlogistics` package carrier type is selected on `product.packaging. This will help greatly the user that must setup those objects.

Here is the list of basic services for Parcel:

```
PostPac Economy: ECO
PostPac Priority: PRI
Bulky goods Economy: SP, ECO
Bulky goods Priority: SP, PRI
PostPac Promo: PPR
PostPac Economy GAS: GAS, ECO
PostPac Priority GAS: GAS, PRI
Bulky goods Economy GAS: GAS, SP, ECO
Bulky good Priority GAS: GAS, SP, PRI
```

PS: Please keep `PostLogistics` brand name with a capital L in UI stings.

simahawk commented on delivery_postlogistics/models/stock_picking.py:
```diff
@@ -0,0 +1,236 @@
+                ("picking_id", "=", self.id),
+            ]
+        )
+        package_ids = []
```
:fire:
```suggestion
        package_ids = set([])
```

sebalix replied:
https://github.com/trobz/delivery-carrier/pull/5

yvaucher replied:
@sebalix please redo your PR on main branch as this was prematurely merged

jbaudoux commented:
> 1. Options are not pulled anymore from the webservice
> 2. Options are new to be set on `product.packaging` inside `shipper_package_code` but there is no guidelines anymore whereas we had the list of option available before.
> 3. As there are additional options, it's not possible to add additional options anymore, we have to select a different `product.packaging`. Which means we have to make almost all possible configuration of packaging? There is no per picking or per package tweaking anymore.

@yvaucher 
Options are defined on the delivery package and on the partner. Both are combined.
We made this change because some options are customer related, and some are related on the carrier/package.

jgrandguillaume commented:
Hi there,

We must merge this work as-is because now this is in production and having it open is hindering our customer when updates are pushed here (long story shot).

I propose to merge this one opened for month now. @phuctranerp make the requested changes by @yvaucher in another PR pleae.

jgrandguillaume commented:
/ocabot merge nobump

sebalix commented:
Fix moved here: https://github.com/OCA/delivery-carrier/pull/356
